Drivers/RcReceiver/Src/SBUSReceiver.cpp

```diff
@@ -0,0 +1,122 @@
+                                            ((raw_sbus_[13] << 8) & 0x07FF));
+        received_sbus_.ch[9]  = static_cast<int16_t>((raw_sbus_[13] >> 3) |
+                                            ((raw_sbus_[14] << 5) & 0x07FF));
+        received_sbus_.ch[10] = static_cast<int16_t>((raw_sbus_[14] >> 6) |
+                                            (raw_sbus_[15] << 2) |
+                                            ((raw_sbus_[16] << 10) & 0x07FF));
+        received_sbus_.ch[11] = static_cast<int16_t>((raw_sbus_[16] >> 1) |
+                                            ((raw_sbus_[17] << 7) & 0x07FF));
+        received_sbus_.ch[12] = static_cast<int16_t>((raw_sbus_[17] >> 4) |
+                                            ((raw_sbus_[18] << 4) & 0x07FF));
+        received_sbus_.ch[13] = static_cast<int16_t>((raw_sbus_[18] >> 7) |
+                                            (raw_sbus_[19] << 1) |
+                                            ((raw_sbus_[20] << 9) & 0x07FF));
+        received_sbus_.ch[14] = static_cast<int16_t>((raw_sbus_[20] >> 2) |
+                                            ((raw_sbus_[21] << 6) & 0x07FF));
+        received_sbus_.ch[15] = static_cast<int16_t>((raw_sbus_[21] >> 5) |
+                                            ((raw_sbus_[22] << 3) & 0x07FF));
+
+        /* CH 17 */
+        received_sbus_.ch17 = raw_sbus_[23] & CH17_MASK_;
+        /* CH 18 */
+        received_sbus_.ch18 = raw_sbus_[23] & CH18_MASK_;
+        /* Grab the lost frame */
+        received_sbus_.lost_frame = raw_sbus_[23] & LOST_FRAME_MASK_;
+        /* Grab the failsafe */
+        received_sbus_.failsafe = raw_sbus_[23] & FAILSAFE_MASK_;
+
+        received_sbus_.new_data = true;
+
+    }else{
+    	received_sbus_.new_data = false;
+    }
+}
+
+void SBUSReceiver::cast_rccontrol()
+{
+    for(uint8_t i = 0; i < 16; i++){
+        received_rccontrol_.ControlSignals[i] = sbus_to_rccontrol(received_sbus_.ch[i]);
+    }
+}
+
+ float SBUSReceiver::sbus_to_rccontrol(uint16_t channel_value)
+ {
+    if(channel_value < SBUS_RANGE_MIN)
+        channel_value = SBUS_RANGE_MIN;
+    if(channel_value > SBUS_RANGE_MAX)
+        channel_value = SBUS_RANGE_MAX;
+    return static_cast<float>((channel_value - SBUS_RANGE_MIN) * (100.0f / SBUS_RANGE_RANGE));
+ }
+
+ void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
+ {
+	 HAL_GPIO_WritePin(GPIOB, GPIO_PIN_7,GPIO_PIN_SET);
+	 sbus_pointer->parse();
+	 HAL_UART_Receive_DMA (huart, sbus_pointer->raw_sbus_, SBUS_FRAME_SIZE);
+//	 SBUSSender::getInstance(huart)->assemble_packet();
+//	 HAL_UART_Transmit(huart, SBUSSender::getInstance(huart)->send_buf_, SBUS_FRAME_SIZE, 10);
+	 HAL_GPIO_WritePin(GPIOB, GPIO_PIN_7,GPIO_PIN_RESET);
+ }
+
+ void HAL_UART_ErrorCallback(UART_HandleTypeDef *huart)
+ {
+	 HAL_GPIO_WritePin(GPIOA, GPIO_PIN_9,GPIO_PIN_SET);
+	 HAL_UART_DMAStop(huart);
+	 HAL_UART_Receive_DMA(huart, sbus_pointer->raw_sbus_, SBUS_FRAME_SIZE);
```

Comment:
The DMA callback will eventually be shared within the whole program, it doesn't only belong to the SBUS driver. I suggest moving the callback function to Common directory, where all the dma usage can be seen there.

Reply:
like having a Drivers/Common folder for this? I like that idea